src/flp/types/fixedpoint_l2.rs

```diff
@@ -0,0 +1,631 @@
+/// type `T`, and the aggregate result is the float vector of the sum of the measurements.
+///
+/// The validity circuit verifies that the L2 norm of each measurement is bounded by 1.
+///
+/// The [*fixed* crate] is used for fixed point numbers, in particular, exactly the following types
+/// are supported: `FixedI8<U7>`, `FixedI16<U15>`, `FixedI32<U31>`, `FixedI64<U63>` and
+/// `FixedI128<U127>`.
+///
+/// Depending on the size of the vector that needs to be transmitted, a corresponding field type has
+/// to be chosen for `F`. For a `n`-bit fixed point type and a `d`-dimensional vector, the field
+/// modulus needs to be larger than `d * 2^(2n-2)` so there are no overflows during norm validity
+/// computation.
+///
+/// [*fixed* crate]: https://crates.io/crates/fixed
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct FixedPointL2BoundedVecSum<T: Fixed, F: FieldElement> {
+    bits_per_entry: usize,
+    entries: usize,
+    bits_for_norm: usize,
+    range_01_checker: Vec<F>,
+    square_computer: Vec<F>,
+    phantom: PhantomData<T>,
+    // range/position constants
+    range_norm_begin: usize,
+    range_norm_end: usize,
+}
+
+/* In the following a high level overview over the inner workings of this type
+ * is given and implementation details are discussed. It is not necessary for
+ * using the type, but it should be very helpful when trying to understand the
+ * implementation.
+ *
+ *
+ * --- Overview ---
+ * Clients submit a vector of numbers whose values semantically lie in [-1,1),
+ * together with a norm in the range [0,1). The validation circuit checks that
+ * the norm of the vector is equal to the submitted norm, while the encoding
+ * guarantees that the submitted norm lies in the correct range.
+ *
+ *
+ * --- Different number encodings ---
+ * Let n denote the number of bits of the chosen fixed-point type.
+ * Numbers occur in 5 different representations:
+ *  (1) Clients have a vector whose entries are fixed point numbers. Only those
+ *      fixed point types are supported where the numbers lie in the range
+ *      [-1,1).
+ *  (2) Because norm computation happens in the validation circuit, it is done
+ *      on entries encoded as field elements. That is, the same vector entries
+ *      are now represented by integers in the range [0,2^n), where -1 is
+ *      represented by 0 and +1 by 2^n.
+ *  (3) Because the field is not necessarily exactly of size 2^n, but might be
+ *      larger, it is not enough to encode a vector entry as in (2) and submit
+ *      it to the aggregator. Instead, in order to make sure that all submitted
+ *      values are in the correct range, they are bit-encoded. (This is the same
+ *      as what happens in the `Sum` type.)
+ *      This means that instead of sending a field element in the range [0,2^n),
+ *      we send n field elements representing the bit encoding. The validation
+ *      circuit can verify that all submitted "bits" are indeed either 0 or 1.
```

Comment:
For a d-sized vector of n-bit fixed point numbers the encoded input + proof is in the order of `d*n` (actually `(d+2)*n - 2`). And since `d` can become very high it seems likely that we will have to do everything we can to keep the communication cost as small as possible, so this idea is very relevant!

If instead of binary encoding, we switch to a b-ary encoding of our vector entries, we could reduce by a factor of `log_2(b)`, while the polynomial in the range-checker gets larger by a factor of `b/2`. I do not know currently how this will affect the circuit (or validation-) runtime, is a basis of b=256 realistic for example?

Anyways, (at least from our perspective) it would be convenient to begin thinking about such optimizations later, when we have a prototype for our full pipeline. Would it be ok to keep the encoding as is for now, and probably submit an update for it later?

Reply:
(Missed this in my second pass, sorry!) Yes, I'd be OK with that. Since we won't be standardizing this any time, it seems reasonable to keep this feature experimental for now. However, one fairly straight forward optimization is outlined here: https://github.com/divviup/libprio-rs/pull/283#discussion_r954458953. It shouldn't be too bad to evaluate quickly, if you have the time.